File: surf-api-core/surf-api-core-api/src/main/java/dev/slne/surf/surfapi/core/api/command/SurfCommandUtil.java
package dev.slne.surf.surfapi.core.api.command;

import static com.google.common.base.Preconditions.checkNotNull;

import dev.jorel.commandapi.exceptions.WrapperCommandSyntaxException;
import dev.slne.surf.surfapi.core.api.command.builder.CommandExceptionBuilder;
import dev.slne.surf.surfapi.core.api.command.exception.WrapperCommandExceptionComponent;
import net.kyori.adventure.text.ComponentLike;
import org.jetbrains.annotations.ApiStatus.NonExtendable;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

@NonExtendable
public interface SurfCommandUtil {

  @Contract("_ -> new")
  static @NotNull WrapperCommandExceptionComponent createException(ComponentLike message) {
    return new WrapperCommandExceptionComponent(checkNotNull(message, "message").asComponent());
  }

  @Contract("_ -> fail")
  static void failWithMessage(ComponentLike message) throws WrapperCommandSyntaxException {
    throw createException(message);
  }

  @Contract("_ -> fail")
  static void failWithBuilder(@NotNull CommandExceptionBuilder builder) throws WrapperCommandSyntaxException {
    throw createException(builder.build());
  }
}

File: surf-api-core/surf-api-core-api/src/main/kotlin/dev/slne/surf/surfapi/core/api/command/SurfCommandUtil.kt
package dev.slne.surf.surfapi.core.api.command

import dev.slne.surf.surfapi.core.api.command.builder.CommandExceptionBuilder
import dev.slne.surf.surfapi.core.api.command.exception.WrapperCommandExceptionComponent
import net.kyori.adventure.text.ComponentLike

/**
 * Utility object for command-related operations and error handling in the Surf API.
 * Provides methods to create and throw custom command exceptions with user-friendly error messages.
 */
object SurfCommandUtil {

    /**
     * Creates a custom exception with the given message.
     *
     * @param message The message to include in the exception, as a [ComponentLike].
     * @return An instance of [WrapperCommandExceptionComponent].
     */
    @JvmStatic
    fun createException(message: ComponentLike) =
        WrapperCommandExceptionComponent(message.asComponent())

    /**
     * Throws a custom command exception with the given message.
     *
     * @param message The message to include in the exception, as a [ComponentLike].
     * @throws WrapperCommandExceptionComponent Always throws the created exception.
     */
    @JvmStatic
    fun failWithMessage(message: ComponentLike) {
        throw createException(message)
    }

    /**
     * Throws a custom command exception using a provided [CommandExceptionBuilder].
     *
     * @param builder The builder used to construct the exception message.
     * @throws WrapperCommandExceptionComponent Always throws the created exception.
     */
    @JvmStatic
    fun failWithBuilder(builder: CommandExceptionBuilder) {
        throw createException(builder.build())
    }
}
